### PR TITLE
docs: added websocket authentication

### PR DESCRIPTION
**Pull request by juzhiyuan**
### Description

This tutorial uses `key-auth` to authenticate before connecting to the upstream WebSocket server.

Rendered Version: https://github.com/apache/apisix/blob/docs/tutorial-websocket-authentication/docs/en/latest/tutorials/websocket-authentication.md

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: docs/en/latest/tutorials/websocket-authentication.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+Connect `ws://127.0.0.1:9080/raw` without `key`, APISIX returns `401 Unauthorized` status code.
+
+![Connect without Key](https://static.apiseven.com/2022/12/06/638ef6db9dd4b.png)
+
+#### Connect with Key
+
+1. Add one header `apikey` with value `this_is_the_key`;
+2. Connect `ws://127.0.0.1:9080/raw` with `key`, it's successfully.
+
+![Connect with key](https://static.apiseven.com/2022/12/06/638efac7c42b6.png)
```
Comment (bzp2010) [inline docs/en/latest/tutorials/websocket-authentication.md:136]: ditto

**Review by bzp2010 — COMMENTED** [review]
The content of the article looks good, but there is a problem here.

AFAIK, the browser's WebSocket does not provide an API to allow users to set their own initial request headers. It will affect browser users, but not users who build their own clients.

So for browsers, we'd better switch to using the query parameter as authentication to start a WS connection, otherwise it will be difficult for users to use the authentication plugin provided by APISIX in the browser.

Perhaps it would be better to provide a truly dedicated websocket plugin in the future. 😊

File: docs/en/latest/tutorials/websocket-authentication.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+#
+-->
+
+Apache APISIX supports [WebSocket](https://en.wikipedia.org/wiki/WebSocket) traffic, but the WebSocket protocol doesn't handle authentication. This article guides you on how to configure authentication for WebSocket connections.
+
+## WebSocket Protocol
+
+To establish a WebSocket connection, the client sends a WebSocket **handshake** request, for which the server returns a WebSocket handshake response, see below:
+
+**Client Request**
```
Comment (navendu-pottekkat) [inline docs/en/latest/tutorials/websocket-authentication.md:36]: Would it be better to change bolds to H3 heading?

File: docs/en/latest/tutorials/websocket-authentication.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+Apache APISIX supports [WebSocket](https://en.wikipedia.org/wiki/WebSocket) traffic, but the WebSocket protocol doesn't handle authentication. This article guides you on how to configure authentication for WebSocket connections.
+
+## WebSocket Protocol
```
Comment (navendu-pottekkat) [inline docs/en/latest/tutorials/websocket-authentication.md:32]: All headings should follow sentence case as stated in the style guide: https://apisix.apache.org/docs/general/documentation-style-guide/#formatting-punctuation-and-organization

File: docs/en/latest/tutorials/websocket-authentication.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+![Websocket Handshake Workflow](https://static.apiseven.com/2022/12/06/638eda2e2415f.png)
+
+## WebSocket Authentication
+
+Apache APISIX supports several ways to do authentication, for example: [basic-auth](https://apisix.apache.org/docs/apisix/plugins/basic-auth/), [key-auth](https://apisix.apache.org/docs/apisix/plugins/key-auth/), [jwt-auth](https://apisix.apache.org/docs/apisix/plugins/jwt-auth/), and so on.
+
+When establishing one connection from Client to Server, in the **handshake** phase, APISIX first checks its authentication information, then chooses to proxy this request or deny it directly.
+
+### Pre-requisite
```
Comment (navendu-pottekkat) [inline docs/en/latest/tutorials/websocket-authentication.md:69]:
```suggestion
### Prerequisites
```

File: docs/en/latest/tutorials/websocket-authentication.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+:::tip
+
+APISIX 3.0 and APISIX 2.x are using different Admin API endpoint. Please check [APISIX 3.0 Deployment Modes](https://apisix.apache.org/docs/apisix/deployment-modes/).
+
+:::
+
+### Key Auth
+
+#### Create one Route
```
Comment (navendu-pottekkat) [inline docs/en/latest/tutorials/websocket-authentication.md:82]:
```suggestion
#### Create a Route
```

File: docs/en/latest/tutorials/websocket-authentication.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        },
+        "scheme": "https"
+    },
+    "plugins": {
+        "key-auth": {}
+    }
+}'
+```
+
+#### Create one Consumer
```
Comment (navendu-pottekkat) [inline docs/en/latest/tutorials/websocket-authentication.md:112]:
```suggestion
#### Create a Consumer
```

File: docs/en/latest/tutorials/websocket-authentication.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+#### Connect with Key
+
+1. Add one header `apikey` with value `this_is_the_key`;
+2. Connect `ws://127.0.0.1:9080/raw` with `key`, it's successfully.
+
+![Connect with key](https://static.apiseven.com/2022/12/06/638efac7c42b6.png)
+
+### Note
+
+Other authentication methods are similar to this one.
```
Comment (navendu-pottekkat) [inline docs/en/latest/tutorials/websocket-authentication.md:143]: Please move this note to the top and use the Docusaurus admonition `:::note`.

File: docs/en/latest/tutorials/websocket-authentication.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+## Reference
+
+1. [Wikipedia - WebSocket](https://en.wikipedia.org/wiki/WebSocket)
```
Comment (navendu-pottekkat) [inline docs/en/latest/tutorials/websocket-authentication.md:147]: We do not generally have separate references section. You can just link it inline as you have done in the first sentence. We can remove this section.

**Review by navendu-pottekkat — CHANGES_REQUESTED** [review]

**Comment by navendu-pottekkat** [comment]
I will update this PR because I think @juzhiyuan is busy elsewhere.

**Comment by navendu-pottekkat** [comment]
Opened new PR to add this tutorial. Closing this one.

See: https://github.com/apache/apisix/pull/9369

**Pull request closed without merging**